### DISCUSSION TITLE
Massive performance issue on very large repo

### DISCUSSION THREAD

**Discussion by HenrikPoulsen** [renovatebot/renovate#20818] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.159.1

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

3.7.6

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Renovate intermittently takes a very long time to run on a large monorepo. 15k branches and many hundred active contributors.
The thing that stands out in the logs is what you see in the debug logs section.
Took a look in the code and that seems to run `git branch --remotes --verbose --contains {sha}`

Running that on my computer with a `time` prepended to it gives the following:
```
real    11m9.551s
user    11m8.929s
sys     0m0.110s
```
So it is extremely slow on that repo.
Is there any other way that code can accomplish what it needs without enumerating all 15k branches?
What are the options available to me to prevent this code to run at all? Is there such a method?

### What you tried so far.

Haven't tried anything other than that I was able to reproduce that it is in fact a very slow command

### Relevant debug logs

<details><summary>Logs</summary>

```
2023-03-09T09:51:32.3977423Z DEBUG: branch.isBehindBase(): using git to calculate (repository=org/repo)
2023-03-09T10:09:58.1194894Z DEBUG: branch.isBehindBase(): false (repository=org/repo)
```

</details>

**Comment by rarkins** [comment]
That particular git command is part of the "is the branch behind base branch?" logic: https://github.com/renovatebot/renovate/blob/a3cc159f7e30120da72b60196f3f8d706d5d5709/lib/util/git/index.ts#L567-L606

It looks like we're not caching the _result_, but even doing so would not be too helpful if it takes 11minutes to get a response.

**Reply by HenrikPoulsen** [reply]
Is there any value in having that be possible to disable entirely? I'm not sure what the value of that logic is. Is it mainly to determine if the PR is obsolete and should be auto closed or something else?
If it's non-critical I would say we would be happy with the option to be able to disable it entirely for that repo if it just means some PRs won't be auto closed or whatever the situation may be.
We have already set `"rebaseWhen": "conflicted"` for example to reduce the activity.

Only other way I can think of to make this work on larger repos is to do this locally in the local clone renovate has. So fetch all the baseBranches, ensure they are fetched in the clone, then run the same command but drop the `--remote`.
Would that introduce other issues? Other than that the branches might have had new commits on them since the clone happened of course, but I imagine you could re-fetch the branches whenever isBranchBehindBase is called, then followed by the git branch command 🤷

**Reply by rarkins** [reply]
In theory if you don't have rebaseWhen=behind-base-branch then Renovate shouldn't need to check it.

**Reply by HenrikPoulsen** [reply]
The debug logs print:
```
Skipping behind base branch check due to rebaseWhen=conflicted
```
So it is correctly skipping in some cases. However there seems to be other calls to that function that do not have that check.

I see you have this: https://github.com/renovatebot/renovate/blob/a72998724ef99bf2e887eedad7dd94c43bf61c83/lib/workers/repository/update/branch/reuse.ts#L54-L58

But there's another flow that gets you in there. Here's the callstack:
```
isBranchBehindBase(), index.ts:571
isBranchBehindBase(), default-scm.ts:23
generateBranchCache(), cache.ts:64
Async call from await
Async call from await
Async call from await
Async call from await
setBranchCache(), cache.ts:107
renovateRepository(), index.ts:86
```
So from here:
https://github.com/renovatebot/renovate/blob/a72998724ef99bf2e887eedad7dd94c43bf61c83/lib/workers/repository/cache.ts#L64
Which does not have that check. Is that a bug then?

**Reply by kwelch** [reply]
We also have this problem and this call alone takes over an hour per branch on some of our repos. Any way I help dig into ways to speed this up.

**Reply by HenrikPoulsen** [reply]
I tried just copying over the if statements from one place to the other location as seen in the image
![image](https://user-images.githubusercontent.com/472627/224167013-0672febd-9a4b-4ee9-adc8-6786c9b25772.png)
Then tried that against the repo in question.
All those 20min chunks vanished, and it went from 1h+ (we our runs time out after 1 hour for technical reasons) down to 13 min total (6 of those is cloning the repo).
So if we could ship some variant of the changes in the image that would essentially solve the issue for my case.
I want to avoid making my own workarounds in the code when possible.
So just let me know if you would need a hand to make a PR out of those changes.

**Reply by rarkins** [reply]
I'll look into it

**Reply by rarkins** [reply]
Can you check out the branch from #20837 and confirm if that also solves you problem?

**Reply by HenrikPoulsen** [reply]
Haven't tried it yet. Don't see how it would work though tbh. `setCachedBehindBaseResult()` isn't called anywhere in the code, as mentioned in the PR.
But in my case I don't want it to be called at all. Since even running it once per PR is still way too much. Is the intent to cache `undefined` when the check is to be skipped?

**Reply by rarkins** [reply]
Yes

**Reply by HenrikPoulsen** [reply]
I can confirm that it does not call the slow command with the branch you mentioned, it sets isBehindBase to undefined and then moves on.
So from that perspective it seems good.
Does the cached value get set somewhere else? Since as I mentioned `setCachedBehindBaseResult()` isn't called anywhere. Without knowing that part of the code it just seems like that is what you would expect. Set is called to cache it, then get to fetch it later. Or is somewhere in the code inserting that directly in the cache without calling that function? If so then I imagine it's fine to ship 🤞

**Reply by rarkins** [reply]
Will fix that first

**Comment by rarkins** [comment]
Related: #20837 and #20839

**Comment by kwelch** [comment]
Note: not an answer. 

 I am just curious why Renovate is using `git branch --remotes --verbose --contains {sha}` this command specifically to determine if the branch is behind. Why are we checking all branches? Shouldn't this just be a check between the update branch and base branch?
 
 I have not have examined this deeply, but we are seeing this error resulting in hours of processing, so I am trying to understand it better so we can help bring that time down.

**Reply by rarkins** [reply]
This is an easy way to know whether a branch is derived from another commit. Even ChatGPT recommended this approach when I asked it today:

<img width="736" alt="image" src="https://user-images.githubusercontent.com/6311784/226174688-54fb03cb-3bee-4719-9a56-a35a6e5cd4e0.png">

**Reply by HonkingGoose** [reply]
Here's the official `git branch` manual: [^manual]

> With `--contains`, shows only the branches that contain the named commit (in other words, the branches whose tip commits are descendants of the named commit), `--no-contains` inverts it.

I also found this summary in the `git branch` manual: [^manual-notes]

> The options `--contains`, `--no-contains`, `--merged` and `--no-merged` serve four related but different purposes:
> 
> - `--contains <commit>` is used to find all branches which will need special attention if `<commit>` were to be rebased or amended, since those branches contain the specified `<commit>`.
> - `--no-contains <commit>` is the inverse of that, i.e. branches that don’t contain the specified `<commit>`.
> - `--merged` is used to find all branches which can be safely deleted, since those branches are fully contained by HEAD.
> - `--no-merged` is used to find branches which are candidates for merging into HEAD, since those branches are not fully contained by HEAD.

[^manual]: https://git-scm.com/docs/git-branch#_description
[^manual-notes]: https://git-scm.com/docs/git-branch#_notes

**Comment by rarkins** [comment, marked as answer]
I have rewritten this function in #21036 to not use the origin and will merge it after more testing

**Reply by kwelch** [reply]
OMG! Huge thanks! I likely need to run more tests since these numbers don't feel accurate, but I did a dry-run against our longest running repository using this patch and our current live version. This new version finished in 47mins, and well the other has been running for 17 hours.

**Reply by rarkins** [reply]
Happy to hear more once you're done testing. I was surprised to _not_ see any performance boost in the hosted app after deploying, in fact performance went down a bit at the time but that can be due to a number of things
